core/src/main/java/org/mule/runtime/core/internal/util/collection/TransformingCollection.java

```diff
@@ -78,18 +69,28 @@ public boolean contains(Object o) {
     readLock.lock();
     try {
       boolean contains = delegate.contains(o);
-      if (!contains && o instanceof Message) {
-        contains = delegate.contains(Result.builder((Message) o));
+      if (!contains && isTargetInstance(o)) {
+        readLock.unlock();
+        writeLock.lock();
+        try {
+          contains = delegate.contains(o);
+          if (!contains) {
+            transformAll();
+          }
+        } finally {
+          readLock.lock();
+          writeLock.unlock();
+        }
       }
-      return contains;
+      return delegate.contains(o);
     } finally {
       readLock.unlock();
     }
   }
 
   @Override
-  public Iterator<Message> iterator() {
-    return new ResultToMessageIterator(delegate.iterator(), cursorProviderFactory, eventContext, originatingLocation);
+  public Iterator<T> iterator() {
+    return new TransformingIterator<>(delegate.iterator(), transformer);
```

Comment:
if `transformAll` was already called, this TransformingIterator isn't needed and adds overhead

Reply:
this applies to all menthods that do the transformation